framework/aster-frame/src/softirq/tasklet.rs

```diff
@@ -0,0 +1,169 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use alloc::sync::Arc;
+use core::{
+    cell::RefCell,
+    sync::atomic::{AtomicBool, Ordering},
+};
+
+use intrusive_collections::{intrusive_adapter, LinkedList, LinkedListAtomicLink};
+
+use super::raise_softirq;
+use crate::{
+    arch::irq::{disable_local, enable_local},
+    cpu_local,
+    sync::SpinLock,
+    CpuLocal,
+};
+
+enum TaskLetState {
+    Free,      // The tasklet is free and can be scheduled to any tasklet list.
+    Scheduled, // The tasklet has been scheduled in a tasklet list.
+    Running,   // The tasklet is running.
+}
+
+/// TaskLet mechanism.
+///
+/// # Overview
+///
+/// `TaskLet` provides an abstraction for deferring execution of a small task that
+/// does not need to sleep and can be executed in the interrupt context. It is primarily
+/// used for tasks that cannot be executed directly in the interrupt handler due to
+/// their potentially longer execution time or non-atomic nature.
+///
+/// # Examples
+///
+/// Users can create a `TaskLet` and schedule it at any place.
+/// ```rust
+/// #use aster_frame::softirq::TaskLet;
+///
+/// #fn my_func() {}
+///
+/// let tasklet = TaskLet::new(Arc::new(my_func));
+/// // This tasklet will be executed in softirq context soon.
+/// schedule_tasklet(&tasklet);
+///
+/// ```
+/// Note that the action function should not involve sleep instructions.
+pub struct TaskLet {
+    /// Current state of this tasklet.
+    state: SpinLock<TaskLetState>,
+    link: LinkedListAtomicLink,
+    /// The function that will be called when execute this tasklet.
+    action: Arc<dyn Fn() + Send + Sync + 'static>,
+    /// If this tasklet is disabled.
+    is_disabled: AtomicBool,
+}
+
+impl TaskLet {
+    /// Create a new `TaskLet` instance with its action.
+    pub fn new<F>(action: Arc<F>) -> Arc<Self>
+    where
+        F: Fn() + Send + Sync + 'static,
+    {
+        Arc::new(Self {
+            state: SpinLock::new(TaskLetState::Free),
+            link: LinkedListAtomicLink::new(),
+            action,
+            is_disabled: AtomicBool::new(false),
+        })
+    }
+
+    fn act(&self) {
+        (self.action)()
+    }
+}
+
+intrusive_adapter!(TaskLetAdapter = Arc<TaskLet>: TaskLet { link: LinkedListAtomicLink });
```

Comment:
>  One thing that I am not sure is if using a macro that contains unsafe code such as intrusive_adapter! conflicts with #[deny(unsafe_code)].

It will conflict with #[deny(unsafe_code)]. One alternative approach is to use `LinkedList` in `alloc` crate. In this way when inserting a tasklet into the list will involve memory allocation. This may affect some performance and address issues with memory allocation failures in the future.

Reply:
Then, we find a reason not to use `#[deny(unsafe_code)]`. Using trusted macros to generate unsafe code in certain patterns that are sure to be safe will be a powerful weapon for us to extend the expressiveness of safe Rust.

On the other hand, if such trusted macros are allowed to exist, how can a static analysis tool can differentiate between the unsafe code generated by trusted macros from those by untrusted macros?

Reply:
> It will conflict with #[deny(unsafe_code)]

I figure it out. The conflict is caused by the usage of `#[forbid(unsafe_code)]` in Asterinas and the usage of `#[allow(unsafe_code)]` in `intrusive_adapter!`.

Currently we use `#[forbid(unsafe_code)]` in asterinas. The difference between `#[forbid(unsafe_code)]` and `#[deny(unsafe_code)]` is that we can use `#[allow(unsafe_code)]` to write unsafe code somewhere with `#[deny(unsafe_code)]` and can not do it with `#[forbid(unsafe_code)]`.

I have tried both `#[forbid(unsafe_code)]` and `#[deny(unsafe_code)]`. Both of them can not detect the unsafe code usage hidden in macros. As a result, one can write arbitrary unsafe code in a macro and use this macro in another crate with `#[forbid(unsafe_code)]` or `#[deny(unsafe_code)]`. This is weird, and I suspect the reason is that the unsafe code might be used in macros provided by the Rust standard library, which they deem to be reliable and want to be applicable in any context, even in cases where something like #[forbid(unsafe_code)] is used. This exactly extends the expressiveness of safe Rust.

Return to the problem in this PR, the conflict is caused by the usage of `#[allow(unsafe_code)]` in `intrusive_adapter!`. In fact if this feature does not be added into `intrusive_adapter!`, this macro can be used in the context with `#[forbid(unsafe_code)]` or `#[deny(unsafe_code)]`. But with this feature, `intrusive_adapter!` can be only used in the context with `#[deny(unsafe_code)]`. I guess the reason is that it considers `#[forbid(unsafe_code)]` to be a stronger restriction that is intended for crates that aim to be entirely with pure safe code, and therefore, it does not wish for this macro to be used in such crates. Consequently, I believe we can change the current usage from `#[forbid(unsafe_code)]` to `#[deny(unsafe_code)]`. Both do not prevent the usage of unsafe hidden within macros, and we are open to using such macros that reliably utilize unsafe code as they can extend the capabilities of Safe Rust.

Reply:
> Consequently, I believe we can change the current usage from #[forbid(unsafe_code)] to #[deny(unsafe_code)]

I agree.

> I have tried both #[forbid(unsafe_code)] and #[deny(unsafe_code)]. Both of them can not detect the unsafe code usage hidden in macros.

I think it is the more interesting problem here. Why is rustc designed to have this loophole of allowing unsafe code in macros? To claim a crate non-TCB for Asterinas, we must rely on a tool, instead of human eyes, to determine if the non-TCB crate contains unsafe code or not, even when the unsafe code is hidden inside macros.